### PR TITLE
[rllib] Implement GAE for PPO

### PR DESCRIPTION
**Pull request by pcmoritz**
This PR adds Generalized Advantage Estimation to the PPO implementation, see https://arxiv.org/abs/1506.02438 and https://arxiv.org/abs/1707.06347

### CODE REVIEW

File: python/ray/rllib/models/model.py [hunk 1]
```diff
@@ -15,12 +15,15 @@ class Model(object):
 
     Attributes:
         inputs (Tensor): The input placeholder for this model.
+        prefix (str): Unique identifier to prevent name clashes in the
```
Comment (ericl) [inline python/ray/rllib/models/model.py]: Hm do you actually need this? I thought name scopes were hierarchical so you can reuse this across multiple distinct scopes.

File: python/ray/rllib/policy_gradient/agent.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 from ray.rllib.policy_gradient.env import BatchedEnv
 from ray.rllib.policy_gradient.loss import ProximalPolicyLoss
 from ray.rllib.policy_gradient.filter import MeanStdFilter
-from ray.rllib.policy_gradient.rollout import rollouts, add_advantage_values
+from ray.rllib.policy_gradient.rollout import rollouts, add_return_values, add_advantage_values
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/agent.py]: line length

File: python/ray/rllib/policy_gradient/loss.py [hunk 3]
```diff
@@ -7,12 +7,15 @@
 
 from ray.rllib.models import ModelCatalog
 
+def huber_loss(x, d=2.0):
+    return tf.where(tf.abs(x) < d, 0.5 * tf.square(x), d*(tf.abs(x) - 0.5*d)) # condition, true, false
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/loss.py]: linting

File: python/ray/rllib/policy_gradient/loss.py [hunk 4]
```diff
@@ -7,12 +7,15 @@
 
 from ray.rllib.models import ModelCatalog
 
+def huber_loss(x, d=2.0):
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/loss.py]: what about just using `tf.losses.huber_loss`?

Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/loss.py]: On the other hand, maybe defining it is actually a bit clearer.

File: python/ray/rllib/policy_gradient/policy_gradient.py [hunk 5]
```diff
@@ -22,6 +23,8 @@
     "gamma": 0.995,
     # Number of steps after which the rollout gets cut
     "horizon": 2000,
+    # If true, use the GAE estimator (with a value function)
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/policy_gradient.py]: Can you rewrite this as 

```
# If true, use the Generalized Advantage Estimator (GAE) with a value function.
```

Otherwise people probably may not know the acronym.

File: python/ray/rllib/policy_gradient/policy_gradient.py [hunk 6]
```diff
@@ -170,26 +179,33 @@ def train(self):
             batch_index = 0
             num_batches = (
                 int(tuples_per_device) // int(model.per_device_batch_size))
-            loss, kl, entropy = [], [], []
+            loss, policyloss, vfloss, kl, entropy = [], [], [], [], []
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/policy_gradient.py]: `policyloss` -> `policy_loss` (including as part of other variable names like `batch_policyloss`)

File: python/ray/rllib/policy_gradient/policy_gradient.py [hunk 7]
```diff
@@ -42,6 +45,8 @@
     "rollout_batchsize": 1,
     # Total SGD batch size across all devices for SGD
     "sgd_batchsize": 128,
+    # Coefficient of the value function loss
+    "vfloss_coeff": 1.0,
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/policy_gradient.py]: Consider `vfloss` -> `value_function_loss` everywhere, or at least `vf_loss`

Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/policy_gradient.py]: Same with other instances of `vf`

File: python/ray/rllib/policy_gradient/policy_gradient.py [hunk 8]
```diff
@@ -147,13 +154,15 @@ def train(self):
                     simple_value=traj_len_mean)])
             file_writer.add_summary(traj_stats, self.global_step)
         self.global_step += 1
-        trajectory["advantages"] = ((trajectory["advantages"] -
-                                     trajectory["advantages"].mean()) /
-                                    trajectory["advantages"].std())
+        if config["use_gae"]:
+            trajectory["tdlambdaret"] = ((trajectory["tdlambdaret"] - trajectory["tdlambdaret"].mean()) / trajectory["tdlambdaret"].std())
+        else:
+            trajectory["returns"] = ((trajectory["returns"] - trajectory["returns"].mean()) / trajectory["returns"].std())
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/policy_gradient.py]: this probably still suffers from #846

Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/policy_gradient.py]: in both branches of the if statement

File: python/ray/rllib/policy_gradient/policy_gradient.py [hunk 9]
```diff
@@ -147,13 +154,15 @@ def train(self):
                     simple_value=traj_len_mean)])
             file_writer.add_summary(traj_stats, self.global_step)
         self.global_step += 1
-        trajectory["advantages"] = ((trajectory["advantages"] -
-                                     trajectory["advantages"].mean()) /
-                                    trajectory["advantages"].std())
+        if config["use_gae"]:
+            trajectory["tdlambdaret"] = ((trajectory["tdlambdaret"] - trajectory["tdlambdaret"].mean()) / trajectory["tdlambdaret"].std())
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/policy_gradient.py]: what about `td_lambda_returns`

File: python/ray/rllib/models/model.py (hunk 1 above)
Comment (pcmoritz) [inline python/ray/rllib/models/model.py]: Haha, you are absolutely right, fixed :)

File: python/ray/rllib/policy_gradient/loss.py [hunk 10]
```diff
@@ -37,13 +49,42 @@ def __init__(
...
-                             feed_dict={self.observations: observations})
+        if config["use_gae"]:
+            self.policy_results = [
+                self.sampler, self.curr_logits, self.value_function]
+        else:
+            self.policy_results = [
+                self.sampler, self.curr_logits, tf.constant("NA")]
+
+    def compute(self, observations):
+            return self.sess.run(self.policy_results,
```
Comment (robertnishihara) [inline python/ray/rllib/policy_gradient/loss.py]: I think this will fail linting. Indent by 4

**Comment by pcmoritz** [comment]
retest this please

**Review by robertnishihara — APPROVED** [review]
LGTM
